### PR TITLE
Initial PJS protos

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR contains the initial PJS protos discussed here: https://www.notion.so/2023-10-03-Pachyderm-Job-System-1a6aca512d494adc837e4632d2127576.

### CODE REVIEW

File: src/pjs/pjs.proto [hunk 1]
```diff
@@ -0,0 +1,184 @@
+syntax = "proto3";
+
+package pjs;
+option go_package = "github.com/pachyderm/pachyderm/src/pjs";
```
[the remark this answers is not in the document]
Comment (jrockway) [reply src/pjs/pjs.proto:4]: Yeah.  pjs.pb.go isn't even checked in, it appears.  There is something I want to look at in the generated files, so we should fix that.

**Review by brendoncarroll — APPROVED** [review]

File: src/pjs/pjs.proto [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+import "google/protobuf/any.proto";
+
+message Job {
+  int64 id = 1;
+}
+
+enum JobState {
+  // UNSPECIFIED means the job state is unspecified.
+  JOB_STATE_UNSPECIFIED = 0;
```
Comment (jrockway) [inline src/pjs/pjs.proto:14]: I don't think these enums need the type name in the value names.   The generated code calls it `JobState_JOB_STATE_UNSPECIFIED`, so it just stutters.

File: src/pjs/pjs.proto [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  // FAILED means that the worker processing the job indicated that it failed.
+  FAILED = 1;	
+  // DISCONNECTED means the worker processing the job disconnected.
+  DISCONNECTED = 2;
+  // CANCELED means the job was canceled.
+  CANCELED = 3;
+}
+
+message JobInfo {
+  Job job = 1;
```
Comment (jrockway) [inline src/pjs/pjs.proto:36]: As mentioned above, we should decide what the "forgot to set it" state is; nil job, or job.job_id = 0, but not both.

File: src/pjs/pjs.proto [hunk 4]
```diff
@@ -0,0 +1,184 @@
+syntax = "proto3";
+
+package pjs;
+option go_package = "github.com/pachyderm/pachyderm/v2/src/pjs";
+
+import "google/protobuf/any.proto";
+
+message Job {
+  int64 id = 1;
+}
```
Comment (jrockway) [inline src/pjs/pjs.proto:10]: I feel like embedded messages with one field have caused a lot of pain over the years.  They provide 2 different invalid states; nil Job, and non-nil Job but id = 0.  We should only have one state for unknown/unset job ids.  In that case, I feel like having `int64 job_id` in each message that references a job would be adequate; 0 indicates "the caller forgot to set it".

If we want to keep the embedded message, we should decide right now what the "unset" state is.  Either this should have a validation rule that id is not equal to 0, or each field that takes a job_id as input should have a Job != nil validation rule (`[(validate.rules).message.required = true];`).

File: src/pjs/pjs.proto [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  FAILED = 1;	
+  // DISCONNECTED means the worker processing the job disconnected.
+  DISCONNECTED = 2;
+  // CANCELED means the job was canceled.
+  CANCELED = 3;
+}
+
+message JobInfo {
+  Job job = 1; 
+  Job parent_job = 2;
```
Comment (jrockway) [inline src/pjs/pjs.proto:37]: I assume that there can be parentless jobs, so this is probably a case where a nil Job is the right identifier.  But we should be sure to reject a zero-valued job in that case.

File: src/pjs/pjs.proto [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  JOB_ERROR_CODE_UNSPECIFIED = 0;
+  // FAILED means that the worker processing the job indicated that it failed.
+  FAILED = 1;	
+  // DISCONNECTED means the worker processing the job disconnected.
+  DISCONNECTED = 2;
+  // CANCELED means the job was canceled.
+  CANCELED = 3;
+}
+
+message JobInfo {
```
Comment (jrockway) [inline src/pjs/pjs.proto:35]: Let's document the messages and the fields.

File: src/pjs/pjs.proto [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  rpc CancelJob(CancelJobRequest) returns (CancelJobResponse) {}
+  // DeleteJob deletes a job.
+  // DeleteJob first cancels the job, then deletes all of the metadata and filesets associated with the job.
+  // A job can only be deleted with the parent job context.
+  rpc DeleteJob(DeleteJobRequest) returns (DeleteJobResponse) {}
+  // ListJob returns a list of jobs and information about each job.
+  // The jobs returned in the list are the child jobs of the provided job.
+  // If no job is provided, the list is the child jobs of the provided job context.
+  // The provided job must be associated with the provided job context or a descendant of the job associated with the provided job context.
+  rpc ListJob(ListJobRequest) returns (stream JobInfo) {}
```
Comment (jrockway) [inline src/pjs/pjs.proto:90]: I think we should avoid this pattern of RPCs returning system objects; ListJob should return ListJobResponse; if for now that only contains a JobInfo, that's fine, but the hook for future extensibility without affecting other uses of JobInfo is always appreciated.

File: src/pjs/pjs.proto [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  // ListQueue returns a list of queues and information about each queue.
+  rpc ListQueue(ListQueueRequest) returns (stream QueueInfo) {}
+  // InspectQueue returns detailed information about a queue.
+  rpc InspectQueue(InspectQueueRequest) returns (QueueInfoDetails) {}
+}
+
+// Job Messages
+
+message CreateJobRequest {
+  string context = 1;
```
Comment (jrockway) [inline src/pjs/pjs.proto:119]: Even after reading the doc, I'm not sure what a context is.  It says it identifies the parent of this job, but parents of jobs are identified as Job objects elsewhere.  Why a string and how does a caller produce a correct string?

File: src/typescript/pjs/pjs.pb.ts
Comment (jrockway) [inline src/typescript/pjs/pjs.pb.ts:1]: Someone should edit .gitattributes to make src/typescript/*/*.pb.ts as generated files.

**Review by jrockway — APPROVED** [review]

File: src/pjs/pjs.proto (hunk 8 above)
Comment (brendoncarroll) [reply src/pjs/pjs.proto:119]: This is a bearer token used to create a job from within another job.

File: src/pjs/pjs.proto (hunk 2 above)
Comment (brendoncarroll) [reply src/pjs/pjs.proto:14]: I don't think you can have two enums with the same value in the same package?  Something about c++.

File: src/pjs/pjs.proto (hunk 8 above)
Comment (bbonenfant) [reply src/pjs/pjs.proto:119]: Could this info be included as comment(s) in this file

**Review by bbonenfant — APPROVED** [review]
No need to hold this up, but it helps downstream clients a lot when all additions to the proto file are documented with comments. All the RPCs have great documentation, but the messages at the bottoms of pjs.proto file are a little bare.

File: src/pjs/pjs.proto (hunk 2 above)
Comment (jrockway) [reply src/pjs/pjs.proto:14]: That seems to be true.

**Pull request merged**